lock-api/src/test/java/com/palantir/lock/client/TimestampCorroboratingTimelockServiceTest.java

```diff
@@ -240,6 +244,45 @@ public void metricsNotRegisteredIfNoViolationsDetected() {
         assertThat(taggedMetricRegistry.getMetrics()).isEmpty();
     }
 
+    @Test
+    public void historyIsCorrectlyMaintained() {
+        when(rawTimelockService.getFreshTimestamps(any())).thenReturn(getFreshTimestampsResponse(1L, 1L));
+
+        timelockService = (TimestampCorroboratingTimelockService) TimestampCorroboratingTimelockService.create(
+                NAMESPACE_1, new DefaultTaggedMetricRegistry(), rawTimelockService);
+
+        getFreshTimestamp();
+        assertThrowsClocksWentBackwardsException(this::getFreshTimestamp);
+
+        List<TimestampBoundsRecord> timestampBoundsRecordHistory = timelockService.getTimestampBoundsRecordHistory();
+        assertThat(timestampBoundsRecordHistory).hasSize(2);
+        assertThat(timestampBoundsRecordHistory).isUnmodifiable();
+
+        TimestampBoundsRecord timestampBoundsRecord =
+                timestampBoundsRecordHistory.get(timestampBoundsRecordHistory.size() - 1);
+        assertThat(timestampBoundsRecord.timestampBounds()).isEqualTo(ImmutableTimestampBounds.of(1L, Long.MIN_VALUE));
+        assertThat(timestampBoundsRecord.lowerFreshTimestamp()).isEqualTo(1L);
+        assertThat(timestampBoundsRecord.operationType()).isEqualTo(OperationType.TIMESTAMP);
+    }
+
+    @Test
+    public void limitedHistoryIsStored() {
+        when(rawTimelockService.getFreshTimestamps(any())).thenReturn(getFreshTimestampsResponse(1L, 1L));
+        timelockService = (TimestampCorroboratingTimelockService) TimestampCorroboratingTimelockService.create(
+                NAMESPACE_1, new DefaultTaggedMetricRegistry(), rawTimelockService);
+        getFreshTimestamp();
+
+        int epsilon = 50;
+        IntStream.range(2, 2 * TimestampCorroboratingTimelockService.HISTORY_RECORD_SIZE)
+                .forEach(index -> {
+                    assertThrowsClocksWentBackwardsException(this::getFreshTimestamp);
+                    int expectedSize = Math.min(index, TimestampCorroboratingTimelockService.HISTORY_RECORD_SIZE);
+                    // the cache may take some time to actually evict excessive records
```

Comment:
It's possible to force the cache to perform cleanup (but you'd need to make the cache itself visible for testing, either directly or via a method).